File: models_matlab/model_RDQ18.m
function output = model_RDQ18(input, params)
%__________________________________________________________________________
%
% This code implements the ODE model for sarcomere activation proposed
% in [1].
%
% Inputs:
%  - input: struct containing the input data
%     - input.times: time instants [s]
%     - input.Ca:    intracellular calcium ions concentration [micro M]
%     - input.SL:    sarcomere length [micro m]
%  - params: struct containing the model parameters, or path of a json file
%
% Outputs:
%  - output: struct containing the output data
%     - output.times: time instants [s]
%     - output.Ca:    intracellular calcium ions concentration [micro M]
%     - output.SL:    sarcomere length [micro m]
%     - output.P:     permissivity [-]
%     - output.Ta:    active tension [kPa]
%
%__________________________________________________________________________
%
% Author: Francesco Regazzoni - MOX, Politecnico di Milano
% Email:  francesco.regazzoni@polimi.it
% Date:   2020
%__________________________________________________________________________
%
% References:
%
% [1] F. Regazzoni, L. Dedè, A. Quarteroni "Active contraction of cardiac
%     cells: a reduced model for sarcomere dynamics with cooperative
%     interactions", Biomechanics and Modeling in Mechanobiology (2018)
%     https://doi.org/10.1007/s10237-018-1049-0
%__________________________________________________________________________

    %% Model parameters
    if nargin == 1
        params = '../params/params_RDQ18.json';
    end
    if ischar(params)
        params = jsondecode(fileread(params));
    end
    LA       = params.geometry.LA;             % [micro m]
    LM       = params.geometry.LM;             % [micro m]
    LB       = params.geometry.LB;             % [micro m]
    n_RU     = params.geometry.n_RU;           % [-]
    Lsmooth  = params.geometry.Lsmooth;        % [micro m]
    Q0       = params.RU_steady_state.Q0;      % [-]
    SLQ      = params.RU_steady_state.SLQ;     % [micro m]
    alphaQ   = params.RU_steady_state.alphaQ;  % [micro m^-1]
    mu       = params.RU_steady_state.mu;      % [-]
    gamma    = params.RU_steady_state.gamma;   % [-]
    Kon      = params.RU_kinetics.Kon;         % [micro M^-1 * s^-1]
    Koff     = params.RU_kinetics.Koff;        % [s^-1]
    Kbasic   = params.RU_kinetics.Kbasic;      % [s^-1]
    TaMax    = params.upscaling.TaMax;         % [kPa]

    %% Numerical parameters
    dt = 2.5e-5;               % [s]
    freqenceRatesUpdate = 10;  % [-]

    %% Initialization
    x_i  = @(i) (LM-LB)*.5 * i/n_RU;
    xAZ  = @(SL) (SL-LB)/2;
    xLA  = @(SL) LA - xAZ(SL) - LB;
    xRA  = @(SL) xAZ(SL) - LA;
    ChiRA=@(SL,i) (x_i(i) <= xRA(SL)) .* exp(-(xRA(SL)-x_i(i)).^2 /Lsmooth^2) + ...
                  (x_i(i) > xRA(SL)) .* (x_i(i) < xAZ(SL)) + ...
                  (x_i(i) >= xAZ(SL)) .* exp(-(x_i(i)-xAZ(SL)).^2 /Lsmooth^2);
    ChiLA=@(SL,i) (x_i(i) <= xLA(SL)) .* exp(-(xLA(SL)-x_i(i)).^2 /Lsmooth^2) + ...
                  (x_i(i) > xLA(SL));
    Q    = @(SL) Q0 - alphaQ*(SLQ-SL)*(SL<SLQ);
    Kpn0 = Kbasic*gamma*gamma;
    Kpn1 = Kbasic*gamma*gamma;
    Knp0 = @(SL) Q(SL)*Kbasic/mu;
    Knp1 = @(SL) Q(SL)*Kbasic;
    K1on = Kon;
    K1off= Koff/mu;
    rates         = zeros(n_RU,4,4,4,4); %PC(i,a,b,c,D) = P((a,D,c)_i^t+dt | (a,b,c)_i^t ) / dt
    ones_414      = ones(n_RU,4,1,4);
    expMat        = repmat(permute([0 0 1 1; 0 0 1 1; 1 1 2 2; 1 1 2 2],[3,1,4,2]),n_RU,1,1,1);
    jMat          = repmat((1:n_RU)',1,4,1,4);
    gammaPlusExp  = gamma.^expMat;
    gammaMinusExp = gamma.^-expMat;
    rates(:,:,2,:,1) = Koff*ones_414;
    rates(:,:,3,:,4) = K1off*ones_414;
    rates(:,:,4,:,1) = Kpn0*gammaMinusExp;
    rates(:,:,3,:,2) = Kpn1*gammaMinusExp;
    P_local = @(xODE)  [sum(sum(sum(xODE(1,3:4,:,:),2),3),4); ... % first unit
                        sum(sum(sum(xODE(:,:,3:4,:),2),3),4); ... % central units
                        sum(sum(sum(xODE(end,:,:,3:4),2),3),4)];  % last unit

    %% Time loop
    times = min(input.times):dt:max(input.times);
    Ca    = interp1(input.times,input.Ca,times);
    SL    = interp1(input.times,input.SL,times);
    nT    = length(times);
    P     = zeros(1,nT);

    time_init = tic();
    fprintf('RDQ18 model. Computing... ')
    for iT = 1:nT
        if mod(iT-1, freqenceRatesUpdate) == 0
            % For performance reasons, transition rates are updated every 10 iterations
            update_rates(Ca(iT),SL(iT));
        end
        if iT == 1
            %x(i,a,b,c) = P((X_{i-1},X_i,X_{i+1}) = (a,b,c) )
            x = zeros(n_RU-2,4,4,4);
            % initial state (everything in state 1, i.e. unbounded & non-permissive)
            x(:,1,1,1) = 1;
        else
            x = x + dt * get_rhs(x);
        end
        P(iT) = mean(P_local(x));
    end
    fprintf('done. Time elapsed: %1.3f s\n', toc(time_init))

    output.times = times;
    output.Ca    = Ca;
    output.SL    = SL;
    output.P     = P;
    output.Ta    = TaMax * P;

    %% Functions definition
    function update_rates(Ca, SL)
        ChiRAmat = ChiRA(SL,jMat);
        ChiLAChiRAgammaPlus = ChiRAmat.*ChiLA(SL,jMat).*gammaPlusExp;
        rates(:,:,1,:,2) = Kon.*ChiRAmat*Ca;
        rates(:,:,2,:,3) = Knp1(SL).*ChiLAChiRAgammaPlus;
        rates(:,:,4,:,3) = K1on.*ChiRAmat*Ca;
        rates(:,:,1,:,4) = Knp0(SL).*ChiLAChiRAgammaPlus;
    end

    function rhs = get_rhs(x)
        x2_L = sum(x(2:end,:,:,:),4);                       % x2_L(i-2,a,b) = P( (X_{i-1},X_i) = (a,b) ) , i = 3:n_RU-1
        x2_R = permute(sum(x(1:end-1,:,:,:),2),[1 3 4 2]);  % x2_R(i-1,b,c) = P( (X_i,X_{i+1}) = (b,c) ) , i = 2:n_RU-2
        PhiC = rates(2:(n_RU-1),:,:,:,:) .* x; % probability fluxes (central units)
        PsiL_squeezed(2:(n_RU-2),:,:,:,:) = permute(sum(PhiC(1:(n_RU-3),:,:,:,:),2),[1,3,4,6,5,2])./ x2_L;
        PsiL_squeezed(isnan(PsiL_squeezed))=0;
        PsiL_squeezed(isinf(PsiL_squeezed))=0;
        PsiL_squeezed(1,:,:,:,:) = permute(rates(1,1,:,:,:),[1,3,4,6,5,2]);
        PhiL = PsiL_squeezed .* x; % probability fluxes (left units)
        PsiR_squeezed(1:(n_RU-3),:,:,:,:) = permute(sum(PhiC(2:(n_RU-2),:,:,:,:),4),[1,6,2,3,5,4]) ./ permute(x2_R,[1,4,2,3]);
        PsiR_squeezed(isnan(PsiR_squeezed))=0;
        PsiR_squeezed(isinf(PsiR_squeezed))=0;
        PsiR_squeezed(n_RU-2,:,:,:,:) = permute(rates(n_RU,:,:,1,:),[1,6,2,3,5,4]);
        PhiR = PsiR_squeezed .* x; % probability fluxes (right units)
        rhs = sum( permute(PhiC,[1,2,5,4,3]) - PhiC + permute(PhiL,[1,5,3,4,2]) - PhiL + permute(PhiR,[1,2,3,5,4]) - PhiR,5);
    end
end

File: models_python/model_RDQ18.py
#!/usr/bin/env python3

# Author: Francesco Regazzoni - MOX, Politecnico di Milano
# Email:  francesco.regazzoni@polimi.it
# Date:   2020

import numpy as np
import json
import time

class model_RDQ18:
    """
    Class implementing the ODE model for sarcomere activation proposed in [1].

    References
    ----------

    [1] F. Regazzoni, L. Dedè, A. Quarteroni "Active contraction of cardiac
        cells: a reduced model for sarcomere dynamics with cooperative
        interactions", Biomechanics and Modeling in Mechanobiology (2018)
        https://doi.org/10.1007/s10237-018-1049-0
    """

    def __init__(self, params = '../params/params_RDQ18.json'):
        """
        Constructor.

        Parameters
        ----------
        params : dictionary containing the model parameters, or path of a json
                 file
        """

        self.model_name = 'RDQ18'

        if isinstance(params, str):
            with open(params) as json_file:
                params = json.load(json_file)

        self.LA       = float(params['geometry']['LA'])             # [micro m]
        self.LM       = float(params['geometry']['LM'])             # [micro m]
        self.LB       = float(params['geometry']['LB'])             # [micro m]
        self.n_RU     =   int(params['geometry']['n_RU'])           # [-]
        self.Lsmooth  = float(params['geometry']['Lsmooth'])        # [micro m]
        self.Q0       = float(params['RU_steady_state']['Q0'])      # [-]
        self.SLQ      = float(params['RU_steady_state']['SLQ'])     # [micro m]
        self.alphaQ   = float(params['RU_steady_state']['alphaQ'])  # [micro m^-1]
        self.mu       = float(params['RU_steady_state']['mu'])      # [-]
        self.gamma    = float(params['RU_steady_state']['gamma'])   # [-]
        self.Kon      = float(params['RU_kinetics']['Kon'])         # [micro M^-1 * s^-1]
        self.Koff     = float(params['RU_kinetics']['Koff'])        # [s^-1]
        self.Kbasic   = float(params['RU_kinetics']['Kbasic'])      # [s^-1]
        self.TaMax    = float(params['upscaling']['TaMax'])         # [kPa]

        # Numerical parameters
        self.dt = 2.5e-5                         # [s]
        self.freqenceRatesUpdate = 10            # [-]

        # Initialization
        self.x_i = lambda i: (self.LM-self.LB)*.5 * i/self.n_RU
        self.xAZ = lambda SL: (SL-self.LB)/2.
        self.xLA = lambda SL: self.LA -self.xAZ(SL) - self.LB
        self.xRA = lambda SL: self.xAZ(SL) - self.LA
        self.ChiRA = lambda SL,i: (self.x_i(i) <= self.xRA(SL)) * np.exp(-(self.xRA(SL)-self.x_i(i))**2 /self.Lsmooth**2) + \
                      (self.x_i(i) > self.xRA(SL)) * (self.x_i(i) < self.xAZ(SL)) + \
                      (self.x_i(i) >= self.xAZ(SL)) * np.exp(-(self.x_i(i)-self.xAZ(SL))**2 /self.Lsmooth**2)
        self.ChiLA = lambda SL,i: (self.x_i(i) <= self.xLA(SL)) * np.exp(-(self.xLA(SL)-self.x_i(i))**2 /self.Lsmooth**2) + \
                      (self.x_i(i) > self.xLA(SL))
        self.Q = lambda SL: self.Q0 - self.alphaQ*(self.SLQ-SL)*(SL<self.SLQ)
        self.Kpn0 = self.Kbasic*self.gamma*self.gamma
        self.Kpn1 = self.Kbasic*self.gamma*self.gamma
        self.Knp0 = lambda SL: self.Q(SL)*self.Kbasic/self.mu
        self.Knp1 = lambda SL: self.Q(SL)*self.Kbasic
        self.K1on = self.Kon
        self.K1off = self.Koff/self.mu

        self.rates = np.zeros((self.n_RU,4,4,4,4)) # rates(i,a,b,c,D) = P((a,D,c)_i^t+dt | (a,b,c)_i^t ) / dt
        self.PhiL = np.zeros((self.n_RU-2,4,4,4,4)) # (i,a,b,c,D)
        self.PhiR = np.zeros((self.n_RU-2,4,4,4,4)) # (i,a,b,c,D)
        self.jMat = np.arange(1,self.n_RU+1)
        self.expMat = np.array([[0, 0, 1, 1],[0, 0, 1, 1],[1, 1, 2, 2],[1, 1, 2, 2]])[None,:,:]
        self.gammaPlusExp = self.gamma**self.expMat
        self.gammaMinusExp = self.gamma**-self.expMat
        self.rates[:,:,1,:,0] = self.Koff
        self.rates[:,:,3,:,0] = self.Kpn0*self.gammaMinusExp
        self.rates[:,:,2,:,1] = self.Kpn1*self.gammaMinusExp
        self.rates[:,:,2,:,3] = self.K1off
        self.P_local = lambda x: np.concatenate((x[0,2:4,:,:][None,:].sum(axis=(1,2,3)),    # first unit
                                                 x[:,:,2:4,:].sum(axis=(1,2,3)),            # central units
                                                 x[-1,:,:,2:4][None,:].sum(axis=(1,2,3))))  # last unit

    def solve(self, inputs):
        """
        Perform a simulation with the model.

        Parameters
        ----------
        inputs: dictionary containing the input data
          - inputs['times']: time instants [s]
          - inputs['Ca']:    intracellular calcium ions concentration [micro M]
          - inputs['SL']:    sarcomere length [micro m]

        Returns
        -------
        output: dictionary containing the output data
          - output['times']: time instants [s]
          - output['Ca']:    intracellular calcium ions concentration [micro M]
          - output['SL']:    sarcomere length [micro m]
          - output['P']:     permissivity [-]
          - output['Ta']:    active tension [kPa]
        """

        times = np.arange(np.min(inputs['times']), np.max(inputs['times']), self.dt)
        Ca    = np.interp(times, inputs['times'], inputs['Ca'])
        SL    = np.interp(times, inputs['times'], inputs['SL'])
        nT    = len(times)
        P     = np.zeros(nT)

        old_settings = np.seterr(invalid='ignore') # ignore divide-by-zero warning
        time_init = time.time();
        print('RDQ18 model. Computing... ', end = '')
        for iT in range(nT):
            if iT % self.freqenceRatesUpdate == 0:
                # For performance reasons, transition rates are updated every 10 iterations
                self.__update_rates(Ca[iT],SL[iT])
            if iT == 0:
                #x(i,a,b,c) = P((X_{i-1},X_i,X_{i+1}) = (a,b,c) )
                x = np.zeros((self.n_RU-2,4,4,4))
                # initial state (everything in state 1, i.e. unbounded & non-permissive)
                x[:,0,0,0] = 1.
            else:
                x = x + self.dt * self.__get_rhs(x)
            P[iT] = np.mean(self.P_local(x))
        print('done. Time elapsed: %1.3f s' % (time.time() - time_init))
        np.seterr(**old_settings)

        output = {}
        output['t']  = times
        output['Ca'] = Ca
        output['SL'] = SL
        output['P']  = P
        output['Ta'] = self.TaMax * P

        return output

    def __update_rates(self, Ca, SL):
        ChiRAmat = self.ChiRA(SL,self.jMat)
        ChiLAChiRAgammaPlus = ChiRAmat[:,None,None]*self.ChiLA(SL,self.jMat)[:,None,None]*self.gammaPlusExp
        self.rates[:,:,0,:,1] = self.Kon*ChiRAmat[:,None,None]*Ca
        self.rates[:,:,1,:,2] = self.Knp1(SL)*ChiLAChiRAgammaPlus
        self.rates[:,:,3,:,2] = self.K1on*ChiRAmat[:,None,None]*Ca
        self.rates[:,:,0,:,3] = self.Knp0(SL)*ChiLAChiRAgammaPlus

    def __get_rhs(self, x):
        x2 = x.sum(axis=3)
        self.PhiC = self.rates[1:-1,:,:,:,:] * x[:,:,:,:,None] # probability fluxes (central units)
        self.PhiL[1:,:,:,:,:] = self.PhiC[:-1,:,:,:,:].sum(axis=1)[:,:,:,None,:] / x2[1:,:,:,None,None]
        self.PhiL[np.logical_or(np.isnan(self.PhiL), np.isinf(self.PhiL))] = .0
        self.PhiL[0,:,:,:,:] = self.rates[0,0,:,:,None,:]
        self.PhiL = self.PhiL * x[:,:,:,:,None] # probability fluxes (left units)
        self.PhiR[:-1,:,:,:,:] = self.PhiC[1:,:,:,:,:].sum(axis=3)[:,None,:,:,:] / x2[1:,None,:,:,None]
        self.PhiR[np.logical_or(np.isnan(self.PhiR), np.isinf(self.PhiR))] = .0
        self.PhiR[-1,:,:,:,:] = self.rates[-1,None,:,:,0,:]
        self.PhiR = self.PhiR * x[:,:,:,:,None] # probability fluxes (right units)
        return ( self.PhiC.swapaxes(2,4) - self.PhiC \
               + self.PhiL.swapaxes(1,4) - self.PhiL \
               + self.PhiR.swapaxes(3,4) - self.PhiR ).sum(axis=4)